File: src/lib/core/agent/llm.ts
import { getLLM } from "../llm";
import type { AgentRunner, AgentRunOptions, AgentRunResult } from "./types";

/**
 * LLM-backed agent runner: a single completion through the active LLMProvider,
 * with no external tools. This is the portable default used for classify/author
 * tasks (orchestrator routing, lesson authoring, path planning) that don't need
 * web/shell access. Tool-dependent work should use a ResearchProvider instead.
 */
export class LLMAgentRunner implements AgentRunner {
  readonly name = "llm";
  readonly supportsTools = false;

  async run(opts: AgentRunOptions): Promise<AgentRunResult> {
    const llm = getLLM();
    if (!llm) {
      return { text: "", ok: false, error: "no_llm_configured" };
    }
    try {
      opts.onStep?.("Thinking…");
      const text = await llm.complete(opts.prompt, {
        system: opts.system,
        maxTokens: 2048,
        temperature: 0.4,
      });
      return { text, ok: !!text, error: text ? undefined : "empty_response" };
    } catch (e) {
      return { text: "", ok: false, error: (e as Error).message };
    }
  }
}

File: worker_py/agent/llm.py
"""LLM-backed agent runner (single completion, no tools). Port of ``agent/llm.ts``."""

from typing import List, Optional

from ..llm import get_llm
from .base import AgentRunner, OnStep, RunResult


class LLMAgentRunner(AgentRunner):
    name = "llm"
    supports_tools = False

    async def run(
        self,
        *,
        prompt: str,
        system: Optional[str] = None,
        skills: Optional[List[str]] = None,
        max_turns: int = 40,
        on_step: Optional[OnStep] = None,
    ) -> RunResult:
        llm = get_llm()
        if not llm:
            return RunResult(text="", ok=False, error="no_llm_configured")
        try:
            if on_step:
                on_step("Thinking…")
            text = await llm.complete(prompt, system=system, max_tokens=2048, temperature=0.4)
            return RunResult(text=text, ok=bool(text), error=None if text else "empty_response")
        except Exception as e:  # noqa: BLE001
            return RunResult(text="", ok=False, error=str(e))
